### PR TITLE
feat: add TargetPO trainer

### PR DESCRIPTION
**Pull request by JeanKaddour**
# What does this PR do?

Adds [Target Policy Optimization (TPO)](https://huggingface.co/papers/2604.06159) as a GRPO-family loss with a thin first-class trainer wrapper.

This PR:
- adds `loss_type="tpo"` to `GRPOTrainer`, using the target `q_i ∝ p_i_old * exp(u_i / eta)` and a sequence-level target-matching loss
- builds TPO targets from old-policy sequence logprobs and TPO-whitened group rewards, matching the reference `tpo_skill` population-standardization behavior
- adds `TPOConfig` and `TPOTrainer` as convenience wrappers with TPO defaults and metadata
- documents the minimal `TPOTrainer` usage, prompt-only dataset format, and paper-index entry
- adds focused tests for TPO config defaults, target construction, TPO score whitening, the `p - q` gradient, and public imports

## Validation

- `uv run --extra test pytest tests/test_tpo_trainer.py`
- `uvx ruff check trl/trainer/grpo_trainer.py tests/test_tpo_trainer.py`
- one-step `TPOTrainer` smoke training with `uv run --extra test python`

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Maintainers familiar with GRPO internals and online trainer APIs would be especially helpful reviewers.

---

> [!NOTE]
> **Medium Risk**
> Adds a new online loss path (`loss_type="tpo"`) and distributed, autograd-aware gathering in the GRPO-family training loop, which can affect training correctness and performance across multi-GPU setups. Risk is mitigated by extensive new unit/distributed tests and input validation, but this still touches core optimization logic.
> 
> **Overview**
> Adds Target Policy Optimization (TPO) support to the GRPO family by introducing `loss_type="tpo"` with new config knobs (`tpo_target_temperature`, `tpo_length_normalize_logps`), target construction from rollout-time sequence logprobs plus whitened group rewards, and a sequence-level cross-entropy training objective (optionally with KL regularization).
> 
> Introduces first-class `TargetPOConfig`/`TargetPOTrainer`, exports them publicly, and provides an example training script. Updates docs to list TargetPO across the trainer index, dataset-format expectations, paper index, and performance/memory guides.
> 
> Adds comprehensive tests covering config validation, target/score math, gradient correctness (including prompt groups spanning distributed ranks via autograd-aware all-gather), and Liger-kernel loss parity where available.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 46d1fd154f5cbd62a4509e1eae63c86694207096. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by kashif** [comment]
unfortunate name for the method, as there is already a TPO method  #5506 perhaps this method can be called `TargetPO` ? or another variation?

**Comment by JeanKaddour** [comment]
Oh, I wasn't aware! Thanks for letting me know, will rename soon
